azure_devops.md

```diff
@@ -0,0 +1,230 @@
+
+[A list of Microsoft provided tasks can be viewed here.](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/?view=azure-pipelines)
+
+Check you're working on the right branch and then commit & push this file.
+
+Moving to the Pipelines view in ADO, we can now register our pipeline; unlike GitHub / GitLab, the pipeline doesn't automatically run just because we created a file with a magic name.
+
+To create your pipeline:
+* Click "Create Pipeline"
+* Select "Azure Repos Git" as your code location.
+* Select your repository
+* It should then locate your pipeline file, and you can click "Run"
+
+After a few seconds, you should see it has paused and a warning appears! Grant the pipeline permission to access the Default agent pool that we assigned it.
+
+Now the pipeline will try to run and we see ... **a failure**!
+
+Click into the job, and you should see an issue like:
+> _No agent found in pool Default which satisfies the specified demands_
+
+Unlike GitHub/GitLab, ADO doesn't automatically offer a free tier for running jobs. They will generally grant access to a single free Microsoft-hosted pipeline if [you apply for it](https://learn.microsoft.com/en-us/azure/devops/pipelines/troubleshooting/troubleshoot-start?view=azure-devops#check-for-available-parallel-jobs), but this can take a few business days to approve so won't work for us here.
+
+> If you're curious why - a number of CI/CD tools have restricted their pipeline free tier over the last few years, seemingly due to a rise in abuse of those pipelines with [people using them to mine cryptocurrencies](https://devblogs.microsoft.com/devops/change-in-azure-pipelines-grant-for-public-projects/)!
+
+We will therefore need to host a runner ourselves to give our pipelines somewhere to execute. Thankfully, we can do that directly on a VM hosted in our Azure cloud too.
+
+### Step 4 - Set up a (self-hosted) Pipeline Runner
+
+> The guidance we'll provide is based on the instructions you can find by visiting:
+> * Project Settings (bottom left of the menu in ADO)
+> * Pipelines -> Agent Pools
+> * Click on the "Default" pool, and "New Agent"
+
+In order to create and register your self hosted agent, we'll need access to:
+* A PAT (Personal Access Token) for the agent to have permission to talk to ADO
+* Your ADO organisation URL
+
+You can generate the PAT by clicking on the User Settings (at the top right, by your initials) and visiting the Personal Access Tokens page. For simplicity here, we'll break best security practice and generate a PAT with "Full Access".
+
+Make a note of the value - if you lose this it can't be recovered, although we can always generate a new one.
+
+Your organisation name will be <https://dev.azure.com/\<YOUR TEMPORARY USERNAME HERE\>> e.g. https://dev.azure.com/clouduserblahblah11 (although you can change this if you want).
```

Comment:
Thought this looked a little tidier.
````suggestion
Your organisation name will be 

```
https://dev.azure.com/<YOUR TEMPORARY USERNAME HERE>
```
 e.g. https://dev.azure.com/clouduserblahblah11 (although you can change this if you want).
````